exodus_lambda/functions/origin_request.py

```diff
@@ -290,52 +267,85 @@ def handler(self, event, context):
+            if object_key == "absent":
+                self.logger.info("Item absent for '%s'", uri)
+                return {"status": "404", "statusDescription": "Not Found"}
 
-                self.logger.info(
-                    "The request value for origin_request end is '%s'",
-                    json.dumps(request, indent=4, sort_keys=True),
-                )
+            # Add custom header containing the original request uri
+            request["headers"]["exodus-original-uri"] = [
+                {"key": "exodus-original-uri", "value": request["uri"]}
+            ]
 
-                return request
-            except Exception as err:
-                self.logger.exception(
-                    "Exception occurred while processing %s",
-                    json.dumps(query_result["Items"][0]),
-                )
+            # Update request uri to point to S3 object key
+            request["uri"] = "/" + object_key
+            content_type = (
+                query_result["Items"][0].get("content_type", {}).get("S")
+            )
+            if not content_type:
+                # return "application/octet-stream" when content_type is empty
+                content_type = "application/octet-stream"
 
-                raise err
-        else:
-            self.logger.info("No item found for '%s'", uri)
+            request["querystring"] = urllib.parse.urlencode(
+                {"response-content-type": content_type}
+            )
+
+            self.logger.info(
+                "The request value for origin_request end is '%s'",
+                json.dumps(request, indent=4, sort_keys=True),
+            )
+
+            return request
+        except Exception as err:
+            self.logger.exception(
+                "Exception occurred while processing %s",
+                json.dumps(query_result["Items"][0]),
+            )
+
+            raise err
+
+    def handler(self, event, context):
+        # pylint: disable=unused-argument
+
+        request = event["Records"][0]["cf"]["request"]
+
+        if request["uri"].startswith("/_/cookie/"):
+            return self.handle_cookie_request(event)
+
+        uri = self.resolve_aliases(request["uri"])
+
+        listing_response = self.handle_listing_request(uri)
+        if listing_response:
+            self.set_cache_control(uri, listing_response)
+            return listing_response
+
+        self.logger.info(
+            "The request value for origin_request beginning is '%s'",
+            json.dumps(request, indent=4, sort_keys=True),
+        )
+        self.logger.info(
+            "The uri value for origin_request beginning is '%s'", uri
+        )
+        table = self.conf["table"]["name"]
 
-            # Report 404 to prevent attempts on S3
-            return {"status": "404", "statusDescription": "Not Found"}
+        # Do not permit clients to explicitly request an index file
+        if not uri.endswith(self.index):
```

Comment:
I think this should be endswith('/' + self.index), right?  Otherwise it's blocking the fetch of all files which end in the configured filename, while the desired behavior is to block just the fetch of files whose name precisely matches the index.